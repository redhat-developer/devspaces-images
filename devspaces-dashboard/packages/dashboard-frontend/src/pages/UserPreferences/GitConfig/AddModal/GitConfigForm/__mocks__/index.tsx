/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/AddModal/GitConfigForm';
import * as GitConfigStore from '@/store/GitConfig';

export class GitConfigForm extends React.PureComponent<Props> {
  render() {
    const { onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { name: 'User One' } } as GitConfigStore.GitConfig, false)
          }
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange({ user: { email: 'user-1@chetest.com', name: 'User One' } }, true)
          }
        />
      </div>
    );
  }
}
