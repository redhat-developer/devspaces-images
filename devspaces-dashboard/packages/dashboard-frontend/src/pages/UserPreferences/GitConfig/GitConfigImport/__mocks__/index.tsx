/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitConfig/GitConfigImport';

export class GitConfigImport extends React.PureComponent<Props> {
  render() {
    const { content, onChange } = this.props;

    return (
      <div>
        <input
          data-testid="submit-invalid-git-config"
          type="button"
          value={content}
          onClick={() => onChange('[user]\n\tname = User One\n', false)}
        />
        <input
          data-testid="submit-valid-git-config"
          type="button"
          onClick={() =>
            onChange('[user]\n\tname = User One\n\temail = user-1@chetest.com\n', true)
          }
        />
      </div>
    );
  }
}
