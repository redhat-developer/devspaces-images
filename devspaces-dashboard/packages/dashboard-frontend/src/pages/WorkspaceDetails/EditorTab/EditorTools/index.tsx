/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Flex, FlexItem, Button, Divider } from '@patternfly/react-core';
import {
  CompressIcon,
  CopyIcon,
  DownloadIcon,
  ExpandIcon,
} from '@patternfly/react-icons/dist/js/icons';
import CopyToClipboard from 'react-copy-to-clipboard';
import stringify from '../../../../services/helpers/editor';

import './EditorTools.styl';

type Props = {
  devfile: che.WorkspaceDevfile;
  handleExpand: (isExpand: boolean) => void;
};

type State = {
  copied?: boolean;
  isExpanded: boolean;
};

class EditorTools extends React.PureComponent<Props, State> {
  private readonly handleExpand: () => void;

  constructor(props: Props) {
    super(props);

    this.state = {
      isExpanded: false,
    };

    this.handleExpand = () => {
      if (this.state.isExpanded) {
        window.postMessage('show-navbar', '*');
        const isExpanded = false;
        this.setState({ isExpanded });
        this.props.handleExpand(isExpanded);
      } else {
        window.postMessage('hide-navbar', '*');
        const isExpanded = true;
        this.setState({ isExpanded });
        this.props.handleExpand(isExpanded);
      }
    };
  }

  public render(): React.ReactElement {
    const devfileText = stringify(this.props.devfile);
    const devfileBlobUrl = URL.createObjectURL(
      new Blob([devfileText], { type: 'application/x-yaml' }),
    );

    let copiedTimer;
    const onCopyToClipboard = () => {
      this.setState({ copied: true });
      if (copiedTimer) {
        clearTimeout(copiedTimer);
      }
      copiedTimer = setTimeout(() => {
        this.setState({ copied: false });
      }, 3000);
    };

    return (
      <div className="editor-tools">
        <Flex>
          <FlexItem>
            <CopyToClipboard text={devfileText} onCopy={() => onCopyToClipboard()}>
              <Button variant="link">
                <CopyIcon />
                {this.state.copied ? 'Copied' : 'Copy to clipboard'}
              </Button>
            </CopyToClipboard>
          </FlexItem>
          <Divider isVertical />
          <FlexItem>
            <a download="devfile.yaml" href={devfileBlobUrl}>
              <DownloadIcon />
              Download
            </a>
          </FlexItem>
          <Divider isVertical />
          <FlexItem>
            <Button variant="link" onClick={() => this.handleExpand()}>
              {this.state.isExpanded ? (
                <React.Fragment>
                  <CompressIcon />
                  Compress
                </React.Fragment>
              ) : (
                <React.Fragment>
                  <ExpandIcon />
                  Expand
                </React.Fragment>
              )}
            </Button>
          </FlexItem>
        </Flex>
      </div>
    );
  }
}

export default EditorTools;
