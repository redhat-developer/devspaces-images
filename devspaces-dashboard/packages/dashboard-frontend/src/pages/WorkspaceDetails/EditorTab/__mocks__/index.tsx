/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import common from '@eclipse-che/common';
import React from 'react';
import { Props, State } from '..';
import devfileApi from '../../../../services/devfileApi';
import { Workspace } from '../../../../services/workspace-adapter';

export default class EditorTab extends React.PureComponent<Props, State> {
  cancelChanges: () => void;

  constructor(props: Props) {
    super(props);

    this.state = {
      devfile: {} as devfileApi.Devfile,
      hasChanges: false,
      isDevfileValid: true,
      hasRequestErrors: false,
      currentRequestError: '',
      isExpanded: false,
      showDevfileV2ConfirmationModal: false,
    };

    this.cancelChanges = () => {
      // no-op
    };
  }

  async handleOnSave(workspace: Workspace) {
    try {
      await this.props.onSave(workspace);
    } catch (e) {
      this.setState({
        currentRequestError: common.helpers.errors.getMessage(e),
      });
    }
  }

  render() {
    return (
      <div>
        Fake Editor Tab
        <button onClick={() => this.handleOnSave(this.props.workspace)}>Save</button>
        <button onClick={() => this.props.onDevWorkspaceWarning()}>Warning</button>
        <span data-testid="current-request-error">{this.state.currentRequestError}</span>
      </div>
    );
  }
}
