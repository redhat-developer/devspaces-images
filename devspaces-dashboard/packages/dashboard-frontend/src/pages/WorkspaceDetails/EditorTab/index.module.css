/*
* Copyright (c) 2018-2021 Red Hat, Inc.
* This program and the accompanying materials are made
* available under the terms of the Eclipse Public License 2.0
* which is available at https://www.eclipse.org/legal/epl-2.0/
*
* SPDX-License-Identifier: EPL-2.0
*
* Contributors:
*   Red Hat, Inc. - initial API and implementation
*/

.cancelButton,
.saveButton {
  margin-left: 5px;
}

.buttonsGroup {
  position: relative;
  top: -31px;
  height: 0;
}

.editorTab {
  width: calc(100% - 10px);
}

.editorTabExpanded {
  position: absolute;
  top: 0;
  right: 0;
  left: 0;
  height: calc(100vh - 55px);
  padding: 0 20px;
  background-color: var(--pf-global--BackgroundColor--100);
}
