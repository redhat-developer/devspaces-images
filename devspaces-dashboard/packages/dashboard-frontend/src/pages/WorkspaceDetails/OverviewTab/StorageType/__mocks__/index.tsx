/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Props, State } from '..';

export default class StorageTypeFormGroup extends React.PureComponent<Props, State> {
  render() {
    return <div>Fake Storage Type Form</div>;
  }
}
