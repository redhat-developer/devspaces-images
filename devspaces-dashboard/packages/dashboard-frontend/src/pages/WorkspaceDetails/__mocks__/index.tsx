/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Props, State } from '..';

export const mockShowAlert = jest.fn();
export class WorkspaceDetails extends React.PureComponent<Props, State> {
  render() {
    if (!this.props.workspace) {
      return <div>Not found</div>;
    }
    const workspace = this.props.workspace;
    return (
      <div className="WorkspaceDetails">
        Workspace Details Page
        <div data-testid="props-is-loading">{this.props.isLoading}</div>
        <div data-testid="props-old-workspace-path">
          {this.props.oldWorkspaceLocation?.pathname}
        </div>
        <div data-testid="props-show-convert-button">
          {this.props.showConvertButton === true ? 'true' : 'false'}
        </div>
        <div data-testid="props-workspace-id">{this.props.workspace.id}</div>
        <div data-testid="props-workspace-name">{this.props.workspace.name}</div>
        <div data-testid="props-workspaces-link">{this.props.workspacesLink}</div>
        <button onClick={() => this.props.onSave(workspace)}>Save</button>
        <button onClick={() => this.props.onConvert(workspace)}>Convert</button>
      </div>
    );
  }
}
