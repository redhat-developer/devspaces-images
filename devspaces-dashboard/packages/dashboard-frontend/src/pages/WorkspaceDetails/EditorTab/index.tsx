/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import {
  Button,
  Text,
  TextContent,
  Alert,
  AlertActionCloseButton,
  AlertVariant,
  AlertGroup,
  Modal,
  ModalVariant,
  Flex,
  FlexItem,
} from '@patternfly/react-core';
import * as lodash from 'lodash';
import { safeLoad } from 'js-yaml';
import common from '@eclipse-che/common';
import DevfileEditor, { DevfileEditor as Editor } from '../../../components/DevfileEditor';
import EditorTools from './EditorTools';
import { constructWorkspace, isCheWorkspace, Workspace } from '../../../services/workspace-adapter';
import devfileApi, { isDevfileV2, isDevWorkspace } from '../../../services/devfileApi';
import {
  DevWorkspaceClient,
  DEVWORKSPACE_NEXT_START_ANNOTATION,
} from '../../../services/workspace-client/devworkspace/devWorkspaceClient';
import { container } from '../../../inversify.config';

import styles from './index.module.css';

export type Props = {
  onSave: (workspace: Workspace) => Promise<void>;
  workspace: Workspace;
  onDevWorkspaceWarning: () => void;
};

export type State = {
  devfile: che.WorkspaceDevfile | devfileApi.Devfile;
  hasChanges: boolean;
  hasRequestErrors: boolean;
  currentRequestError: string;
  isDevfileValid: boolean;
  isExpanded: boolean;
  copied?: boolean;
  showDevfileV2ConfirmationModal: boolean;
  additionSchema?: { [key: string]: any };
};

export class EditorTab extends React.PureComponent<Props, State> {
  private originDevfile: che.WorkspaceDevfile | devfileApi.Devfile;
  private readonly devfileEditorRef: React.RefObject<Editor>;
  private devworkspaceClient: DevWorkspaceClient;

  cancelChanges: () => void;

  constructor(props: Props) {
    super(props);
    this.devworkspaceClient = container.get(DevWorkspaceClient);

    const devfile = Object.assign({}, this.props.workspace.devfile);
    const additionSchema = this.getAdditionSchema(devfile);

    this.state = {
      devfile,
      hasChanges: false,
      isDevfileValid: true,
      hasRequestErrors: false,
      currentRequestError: '',
      isExpanded: false,
      showDevfileV2ConfirmationModal: false,
      additionSchema,
    };

    this.cancelChanges = (): void => {
      this.updateEditor(this.props.workspace.devfile);
      this.setState({
        hasChanges: false,
        hasRequestErrors: false,
        currentRequestError: '',
      });
    };

    this.devfileEditorRef = React.createRef<Editor>();
  }

  private getAdditionSchema(
    devfile: che.WorkspaceDevfile | devfileApi.Devfile,
  ): { [key: string]: any } | undefined {
    return isDevfileV2(devfile)
      ? {
          properties: {
            metadata: {
              properties: {
                name: {
                  const: (devfile as devfileApi.Devfile).metadata.name,
                },
                namespace: {
                  const: (devfile as devfileApi.Devfile).metadata.namespace,
                },
              },
            },
          },
        }
      : undefined;
  }

  private init(): void {
    const devfile = Object.assign({}, this.props.workspace.devfile);
    if (devfile && (!this.originDevfile || !this.areEqual(devfile, this.originDevfile))) {
      this.originDevfile = devfile;
      this.updateEditor(devfile);
      const additionSchema = this.getAdditionSchema(devfile);
      this.setState({
        hasRequestErrors: false,
        currentRequestError: '',
        hasChanges: false,
        additionSchema,
      });
    }
  }

  public componentDidMount(): void {
    this.init();
  }

  public componentDidUpdate(): void {
    this.init();
  }

  public render(): React.ReactElement {
    const originDevfile = this.props.workspace.devfile;
    const { devfile, additionSchema } = this.state;
    const isReadonly = this.props.workspace.isDeprecated;

    return (
      <React.Fragment>
        <br />
        {this.state.currentRequestError && (
          <Alert
            variant={AlertVariant.danger}
            isInline
            title={this.state.currentRequestError}
            actionClose={
              <AlertActionCloseButton onClose={() => this.setState({ currentRequestError: '' })} />
            }
          />
        )}
        {this.state.showDevfileV2ConfirmationModal && (
          <Modal
            variant={ModalVariant.small}
            isOpen={true}
            title="Restart Workspace"
            onClose={() => this.devfileV2ConfirmationCancellation()}
            actions={[
              <Button key="yes" variant="primary" onClick={() => this.saveDevfile()}>
                Yes
              </Button>,
              <Button
                key="no"
                variant="secondary"
                onClick={() => this.devfileV2ConfirmationCancellation()}
              >
                No
              </Button>,
            ]}
          >
            <TextContent>
              <Text>Would you like to restart the workspace with the changes?</Text>
            </TextContent>
          </Modal>
        )}
        <TextContent
          className={this.state.isExpanded ? styles.editorTabExpanded : styles.editorTab}
        >
          {this.state.currentRequestError && this.state.isExpanded && (
            <AlertGroup isToast>
              <Alert
                variant={AlertVariant.danger}
                title={this.state.currentRequestError}
                actionClose={
                  <AlertActionCloseButton
                    onClose={() => this.setState({ currentRequestError: '' })}
                  />
                }
              />
            </AlertGroup>
          )}
          <EditorTools
            devfile={devfile as che.WorkspaceDevfile}
            handleExpand={isExpanded => {
              this.setState({ isExpanded });
            }}
          />
          <DevfileEditor
            additionSchema={additionSchema}
            ref={this.devfileEditorRef}
            devfile={originDevfile}
            decorationPattern="location[ \t]*(.*)[ \t]*$"
            onChange={(newValue, isValid) => this.onDevfileChange(newValue, isValid)}
            isReadonly={isReadonly}
          />
          {isReadonly === false && (
            <Flex direction={{ default: 'column' }}>
              <FlexItem align={{ default: 'alignRight' }} className={styles.buttonsGroup}>
                <Button
                  onClick={() => this.onSave()}
                  variant="primary"
                  className={styles.saveButton}
                  isDisabled={!this.state.hasChanges || !this.state.isDevfileValid}
                >
                  Save
                </Button>
                <Button
                  onClick={() => this.cancelChanges()}
                  variant="secondary"
                  className={styles.cancelButton}
                  isDisabled={!this.state.hasChanges && this.state.isDevfileValid}
                >
                  Cancel
                </Button>
              </FlexItem>
            </Flex>
          )}
        </TextContent>
      </React.Fragment>
    );
  }

  /**
   * When a devfile v2 user does not allow the devworkspace to restart then store the configuration
   * in an annotation that will be used on next start
   */
  private async devfileV2ConfirmationCancellation() {
    const devfile = this.state.devfile as devfileApi.Devfile;
    if (!devfile) {
      return;
    }
    try {
      await this.checkForModifiedClusterDevWorkspace();
      const devworkspace = this.props.workspace.ref as devfileApi.DevWorkspace;
      const convertedDevWorkspace = constructWorkspace(devworkspace);
      convertedDevWorkspace.devfile = devfile;
      // Store the devfile in here
      (convertedDevWorkspace.ref as devfileApi.DevWorkspace).metadata.annotations = {
        [DEVWORKSPACE_NEXT_START_ANNOTATION]: JSON.stringify(
          convertedDevWorkspace.ref as devfileApi.DevWorkspace,
        ),
      };
      convertedDevWorkspace.ref.status = devworkspace.status;
      this.props.onDevWorkspaceWarning();
      this.props.onSave(convertedDevWorkspace);
      this.setState({
        showDevfileV2ConfirmationModal: false,
      });
    } catch (e) {
      const errorMessage = common.helpers.errors.getMessage(e);
      this.setState({
        hasChanges: true,
        hasRequestErrors: true,
        currentRequestError: errorMessage,
      });
    }
  }

  private updateEditor(devfile: che.WorkspaceDevfile | devfileApi.Devfile): void {
    if (!devfile) {
      return;
    }
    this.devfileEditorRef.current?.updateContent(devfile);
    this.setState({ isDevfileValid: true });
  }

  private onDevfileChange(newValue: string, isValid: boolean): void {
    this.setState({ isDevfileValid: isValid });
    if (!isValid) {
      this.setState({ hasChanges: false });
      return;
    }
    let devfile: che.WorkspaceDevfile;
    try {
      devfile = safeLoad(newValue) as che.WorkspaceDevfile;
    } catch (e) {
      console.error('Devfile parse error', e);
      return;
    }
    if (this.areEqual(this.props.workspace.devfile as che.WorkspaceDevfile, devfile)) {
      this.setState({ hasChanges: false });
      return;
    }
    this.setState({
      devfile,
      hasChanges: true,
      hasRequestErrors: false,
    });
  }

  private async onSave(): Promise<void> {
    if (!this.props.workspace.isRunning || isCheWorkspace(this.props.workspace.ref)) {
      await this.saveDevfile();
    } else {
      this.setState({
        showDevfileV2ConfirmationModal: true,
      });
    }
  }

  /**
   * Check to see if the current devworkspaces devfile and the cluster devworkspaces devfile are the same. If they
   * are not then throw an error
   */
  private async checkForModifiedClusterDevWorkspace(): Promise<void> {
    const currentDevWorkspace = this.props.workspace.ref as devfileApi.DevWorkspace;
    const clusterDevWorkspace = await this.devworkspaceClient.getWorkspaceByName(
      currentDevWorkspace.metadata.namespace,
      currentDevWorkspace.metadata.name,
    );
    if (!lodash.isEqual(clusterDevWorkspace.spec.template, currentDevWorkspace.spec.template)) {
      throw new Error(
        'Could not save devfile to cluster. The clusters devfile and the incoming devfile are different. Please reload the page to get an updated devfile.',
      );
    }
  }

  private async saveDevfile() {
    const devfile = this.state.devfile;
    if (!devfile) {
      return;
    }
    const workspaceCopy = constructWorkspace(this.props.workspace.ref);
    if (!devfile.metadata) {
      devfile.metadata = {};
    }
    if (!devfile.metadata.name) {
      devfile.metadata.name = workspaceCopy.name;
    }
    if (isDevfileV2(devfile) && !devfile.metadata.namespace) {
      devfile.metadata.namespace = workspaceCopy.namespace;
    }

    workspaceCopy.devfile = devfile;
    this.setState({ hasChanges: false });
    try {
      if (isDevWorkspace(workspaceCopy.ref)) {
        await this.checkForModifiedClusterDevWorkspace();
        // We need to manually re-attach devworkspace id so that we can re-use it to re-add default plugins to the devworkspace custom resource
        const dw = this.props.workspace.ref as devfileApi.DevWorkspace;
        workspaceCopy.ref.status = dw.status;
      }
      await this.props.onSave(workspaceCopy);
    } catch (e) {
      const error = common.helpers.errors.getMessage(e).replace(/^Error: /gi, '');
      this.setState({
        hasChanges: true,
        hasRequestErrors: true,
        currentRequestError: error,
      });
    }
    this.setState({
      showDevfileV2ConfirmationModal: false,
    });
  }

  private sortKeysInObject(
    obj: che.WorkspaceDevfile | devfileApi.Devfile,
  ): che.WorkspaceDevfile | devfileApi.Devfile {
    return Object.keys(obj)
      .sort()
      .reduce((result: che.WorkspaceDevfile | devfileApi.Devfile, key: string) => {
        result[key] = obj[key];
        return result;
      }, {} as che.WorkspaceDevfile | devfileApi.Devfile);
  }

  private areEqual(
    a: che.WorkspaceDevfile | devfileApi.Devfile,
    b: che.WorkspaceDevfile | devfileApi.Devfile,
  ): boolean {
    return (
      JSON.stringify(this.sortKeysInObject(a)) ==
      JSON.stringify(this.sortKeysInObject(b as che.WorkspaceDevfile))
    );
  }
}

export default EditorTab;
