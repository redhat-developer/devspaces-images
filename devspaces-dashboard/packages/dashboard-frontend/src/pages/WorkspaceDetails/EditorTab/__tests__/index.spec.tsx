/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { Provider } from 'react-redux';
import { Store } from 'redux';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { dump } from 'js-yaml';
import devfileApi from '../../../../services/devfileApi';
import EditorTab from '..';
import { Workspace, constructWorkspace } from '../../../../services/workspace-adapter';
import { CheWorkspaceBuilder } from '../../../../store/__mocks__/cheWorkspaceBuilder';
import { DevWorkspaceBuilder } from '../../../../store/__mocks__/devWorkspaceBuilder';
import { FakeStoreBuilder } from '../../../../store/__mocks__/storeBuilder';
import { DevWorkspaceClient } from '../../../../services/workspace-client/devworkspace/devWorkspaceClient';
import { container } from '../../../../inversify.config';

// uses the Devfile Editor mock
jest.mock('../../../../components/DevfileEditor');
jest.mock('../EditorTools', () => {
  const FakeEditorTools = () => <div>Editor Tools</div>;
  FakeEditorTools.displayName = 'EditorTools';
  return FakeEditorTools;
});

describe('Editor Tab', () => {
  const mockOnSave = jest.fn();
  const mockOnWorkspaceWarning = jest.fn();

  function getComponent(store: Store, workspace: Workspace): React.ReactElement {
    return (
      <Provider store={store}>
        <EditorTab
          workspace={workspace}
          onSave={workspace => mockOnSave(workspace)}
          onDevWorkspaceWarning={() => mockOnWorkspaceWarning()}
        />
      </Provider>
    );
  }

  afterEach(() => {
    jest.resetAllMocks();
  });

  describe('Che workspaces', () => {
    const workspaceName = 'test-workspace';
    let workspace: Workspace;
    let component: React.ReactElement;

    beforeEach(() => {
      const cheWorkspace = new CheWorkspaceBuilder().withName(workspaceName).build();
      const store = new FakeStoreBuilder()
        .withCheWorkspaces({
          workspaces: [cheWorkspace],
        })
        .build();
      workspace = constructWorkspace(cheWorkspace);
      component = getComponent(store, workspace);
    });

    it('should render the component', () => {
      const json = renderer.create(component).toJSON();
      expect(json).toMatchSnapshot();
    });

    it('should restore name', async () => {
      render(component);

      // copy the workspace devfile, remove 'name' field and paste new devfile into the editor
      const noNameDevfile = JSON.parse(JSON.stringify(workspace.devfile));
      delete noNameDevfile.metadata.name;
      const noNameDevfileContent = dump(noNameDevfile);

      const editor = screen.getByRole('textbox');
      fireEvent.input(editor, {
        target: {
          value: noNameDevfileContent,
        },
      });

      const saveButton = screen.getByRole('button', { name: 'Save' });

      await waitFor(() => expect(saveButton).toBeEnabled());
      userEvent.click(saveButton);

      await waitFor(() => expect(mockOnSave).toHaveBeenCalled());
      expect(mockOnSave).toHaveBeenCalledWith(
        expect.objectContaining({
          workspace: expect.objectContaining({
            devfile: expect.objectContaining({
              metadata: {
                name: workspaceName,
              },
            }),
          }),
        }),
      );
    });
  });

  describe('devworkspaces', () => {
    const workspaceName = 'test-workspace';
    const namespace = 'test-ns';
    let workspace: Workspace;
    let component: React.ReactElement;

    beforeEach(() => {
      const devWorkspace = new DevWorkspaceBuilder()
        .withName(workspaceName)
        .withNamespace(namespace)
        .build();

      const devWorkspaceCopy = JSON.parse(JSON.stringify(devWorkspace));
      // mock devworkspace method to be able to save the devfile
      class MockDevWorkspaceClient extends DevWorkspaceClient {
        /* eslint-disable @typescript-eslint/no-unused-vars */
        async getWorkspaceByName(
          namespace: string,
          workspaceName: string,
        ): Promise<devfileApi.DevWorkspace> {
          return devWorkspaceCopy;
        }
        /* eslint-enable @typescript-eslint/no-unused-vars */
      }
      container.rebind(DevWorkspaceClient).to(MockDevWorkspaceClient).inSingletonScope();

      const store = new FakeStoreBuilder()
        .withDevWorkspaces({
          workspaces: [devWorkspace],
        })
        .build();
      workspace = constructWorkspace(devWorkspace);
      component = getComponent(store, workspace);
    });

    it('should render the component', () => {
      const json = renderer.create(component).toJSON();
      expect(json).toMatchSnapshot();
    });

    it('should restore name', async () => {
      render(component);

      // copy the workspace devfile, remove 'name' field and paste new devfile into the editor
      const noNameDevfile = JSON.parse(JSON.stringify(workspace.devfile)) as devfileApi.Devfile;
      delete (noNameDevfile as devfileApi.DevfileLike).metadata?.name;
      const noNameDevfileContent = dump(noNameDevfile);

      const editor = screen.getByRole('textbox');
      fireEvent.input(editor, {
        target: {
          value: noNameDevfileContent,
        },
      });

      const saveButton = screen.getByRole('button', { name: 'Save' });

      await waitFor(() => expect(saveButton).toBeEnabled());
      userEvent.click(saveButton);

      await waitFor(() => expect(mockOnSave).toHaveBeenCalled());
      expect(mockOnSave).toHaveBeenCalledWith(
        expect.objectContaining({
          workspace: expect.objectContaining({
            kind: 'DevWorkspace',
            metadata: expect.objectContaining({
              name: workspaceName,
            }),
          } as devfileApi.DevWorkspace),
        }),
      );
    });

    it('should restore namespace', async () => {
      render(component);

      // copy the workspace devfile, remove 'name' field and paste new devfile into the editor
      const noNamespaceDevfile = JSON.parse(
        JSON.stringify(workspace.devfile),
      ) as devfileApi.Devfile;
      delete (noNamespaceDevfile as devfileApi.DevfileLike).metadata?.namespace;
      const noNamespaceDevfileContent = dump(noNamespaceDevfile);

      const editor = screen.getByRole('textbox');
      fireEvent.input(editor, {
        target: {
          value: noNamespaceDevfileContent,
        },
      });

      const saveButton = screen.getByRole('button', { name: 'Save' });

      await waitFor(() => expect(saveButton).toBeEnabled());
      userEvent.click(saveButton);

      await waitFor(() => expect(mockOnSave).toHaveBeenCalled());
      expect(mockOnSave).toHaveBeenCalledWith(
        expect.objectContaining({
          workspace: expect.objectContaining({
            kind: 'DevWorkspace',
            metadata: expect.objectContaining({
              namespace,
            }),
          } as devfileApi.DevWorkspace),
        }),
      );
    });
  });
});
