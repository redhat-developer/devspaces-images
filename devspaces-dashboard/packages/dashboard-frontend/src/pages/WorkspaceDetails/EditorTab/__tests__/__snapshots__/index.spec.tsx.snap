// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Tab Che workspaces should render the component 1`] = `
Array [
  <br />,
  "",
  <div
    className="pf-c-content"
  >
    
    <div>
      Editor Tools
    </div>
    <div
      className="devfile-editor"
    >
      <div
        className="monaco"
      >
        <textarea
          id="devfile-editor"
          onChange={[Function]}
          onPaste={[Function]}
          value="apiVersion: 1.0.0
metadata:
  name: test-workspace
"
        />
      </div>
      <div
        className="error"
      />
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        Devfile Documentation
      </a>
    </div>
    <div
      className="pf-l-flex pf-m-column"
    >
      <div
        className="pf-m-align-right"
      >
        <button
          aria-disabled={true}
          aria-label={null}
          className="pf-c-button pf-m-primary pf-m-disabled"
          data-ouia-component-id="OUIA-Generated-Button-primary-1"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={true}
          onClick={[Function]}
          role={null}
          tabIndex={null}
          type="button"
        >
          Save
        </button>
        <button
          aria-disabled={true}
          aria-label={null}
          className="pf-c-button pf-m-secondary pf-m-disabled"
          data-ouia-component-id="OUIA-Generated-Button-secondary-1"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={true}
          onClick={[Function]}
          role={null}
          tabIndex={null}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>,
]
`;

exports[`Editor Tab devworkspaces should render the component 1`] = `
Array [
  <br />,
  "",
  <div
    className="pf-c-content"
  >
    
    <div>
      Editor Tools
    </div>
    <div
      className="devfile-editor"
    >
      <div
        className="monaco"
      >
        <textarea
          id="devfile-editor"
          onChange={[Function]}
          onPaste={[Function]}
          value="schemaVersion: 2.1.0
metadata:
  name: test-workspace
  namespace: test-ns
components: []
"
        />
      </div>
      <div
        className="error"
      />
      <a
        rel="noopener noreferrer"
        target="_blank"
      >
        Devfile Documentation
      </a>
    </div>
    <div
      className="pf-l-flex pf-m-column"
    >
      <div
        className="pf-m-align-right"
      >
        <button
          aria-disabled={true}
          aria-label={null}
          className="pf-c-button pf-m-primary pf-m-disabled"
          data-ouia-component-id="OUIA-Generated-Button-primary-3"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={true}
          onClick={[Function]}
          role={null}
          tabIndex={null}
          type="button"
        >
          Save
        </button>
        <button
          aria-disabled={true}
          aria-label={null}
          className="pf-c-button pf-m-secondary pf-m-disabled"
          data-ouia-component-id="OUIA-Generated-Button-secondary-3"
          data-ouia-component-type="PF4/Button"
          data-ouia-safe={true}
          disabled={true}
          onClick={[Function]}
          role={null}
          tabIndex={null}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>,
]
`;
